pkg/sql/lease.go

```diff
@@ -697,20 +698,38 @@ func (t *tableState) findForTimestamp(
 //
 // t.mu must be locked.
 func (t *tableState) acquireFreshestFromStoreLocked(ctx context.Context, m *LeaseManager) error {
-	// Ensure there is no lease acquisition in progress.
-	t.acquireWait()
-
-	// Move forward to acquire a fresh table lease.
+	// Continue to attempt to call to acquireNodeLease until our call was the only
+	// one made by singleflight. This busy-waits while other acquisitions are
+	// underway.
+	for {
+		// Move forward the expiry to acquire a fresh table lease.
+
+		// Set the min expiration time to guarantee that the lease acquired is the
+		// last lease in t.mu.active .
+		// TODO(vivek): the expiration time is no longer needed to sort the
+		// tableVersionState. Get rid of this.
+		minExpirationTime := hlc.Timestamp{}
+		newestTable := t.mu.active.findNewest()
+		if newestTable != nil {
+			minExpirationTime = newestTable.expiration.Add(int64(time.Millisecond), 0)
+		}
 
-	// Set the min expiration time to guarantee that the lease acquired is the
-	// last lease in t.mu.active .
-	minExpirationTime := hlc.Timestamp{}
-	newestTable := t.mu.active.findNewest()
-	if newestTable != nil {
-		minExpirationTime = newestTable.expiration.Add(int64(time.Millisecond), 0)
+		t.mu.Unlock()
+		// hasAcquiredFreshest is used to determine if our function was the one
+		// called by group.Do. The shared bool returned by Do is not suffice as it
+		// returns true if any other calls to Do use the result, not if our function
+		// returns. If two routines simultaneously call Do here, they endlessly
+		// retry.
+		hasAcquiredFreshest := false
+		_, _, err := t.mu.group.Do("acquire", func() (interface{}, error) {
+			hasAcquiredFreshest = true
+			return nil, t.acquireNodeLease(ctx, m, minExpirationTime)
+		})
+		t.mu.Lock()
```

Comment:
before t.mu.Lock() is acquired t.mu.active.findNewest() can become nil again so we need to call
acquireFreshestFromStoreLocked() in a loop in acquireFreshestFromStore()

Reply:
Yep.

(Hmm, this is a common problem with all uses of `acquireNodeLease` now that it's async and we're handing off the lock. I don't know if there's a better approach to this, but I don't know of one off the top of my head.)

Reply:
Regardless, I've added two tests which catch this problem for both `acquireFreshestFromStore` and `AcquireByName`/`acquire` so I'm a lot more confident in that. The race happens ~50% of the time during `testrace` locally.

Aside: How does teamcity compare to tests locally, when it comes to encountering race conditions? Is it more frequent because of some property of the execution speed?